src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLLayer.m

```diff
@@ -0,0 +1,272 @@
+
+- (id) initWithJavaLayer:(jobject)layer
+{
+    AWT_ASSERT_APPKIT_THREAD;
+    // Initialize ourselves
+    self = [super init];
+    if (self == nil) return self;
+
+    self.javaLayer = layer;
+
+    self.contentsGravity = kCAGravityTopLeft;
+
+    //Disable CALayer's default animation
+    NSMutableDictionary * actions = [[NSMutableDictionary alloc] initWithObjectsAndKeys:
+                                    [NSNull null], @"anchorPoint",
+                                    [NSNull null], @"bounds",
+                                    [NSNull null], @"contents",
+                                    [NSNull null], @"contentsScale",
+                                    [NSNull null], @"onOrderIn",
+                                    [NSNull null], @"onOrderOut",
+                                    [NSNull null], @"position",
+                                    [NSNull null], @"sublayers",
+                                    nil];
+    self.actions = actions;
+    [actions release];
+    self.topInset = 0;
+    self.leftInset = 0;
+    self.framebufferOnly = NO;
+    self.nextDrawableCount = 0;
+    self.opaque = FALSE;
+    CVDisplayLinkCreateWithActiveCGDisplays(&displayLink);
+    CVDisplayLinkSetOutputCallback(displayLink, &displayLinkCallback, (__bridge void*)self);
+    return self;
+}
+
+- (void) blitTexture {
+    if (self.ctx == NULL || self.javaLayer == NULL || self.buffer == nil || self.ctx.device == nil) {
+        J2dTraceLn4(J2D_TRACE_VERBOSE, "MTLLayer.blitTexture: uninitialized (mtlc=%p, javaLayer=%p, buffer=%p, devide=%p)", self.ctx, self.javaLayer, self.buffer, ctx.device);
+        return;
+    }
+
+    if (self.nextDrawableCount != 0) {
+        return;
+    }
+    @autoreleasepool {
+        if ((self.buffer.width == 0) || (self.buffer.height == 0)) {
+            J2dTraceLn(J2D_TRACE_VERBOSE, "MTLLayer.blitTexture: cannot create drawable of size 0");
+            return;
+        }
+        id<MTLCommandBuffer> commandBuf = [self.ctx createBlitCommandBuffer];
+        if (commandBuf == nil) {
+            J2dTraceLn(J2D_TRACE_VERBOSE, "MTLLayer.blitTexture: commandBuf is null");
+            return;
+        }
+        id<CAMetalDrawable> mtlDrawable = [self nextDrawable];
+        if (mtlDrawable == nil) {
+            J2dTraceLn(J2D_TRACE_VERBOSE, "MTLLayer.blitTexture: nextDrawable is null)");
+            return;
+        }
+        self.nextDrawableCount++;
+        id <MTLBlitCommandEncoder> blitEncoder = [commandBuf blitCommandEncoder];
+
+        [blitEncoder
+                copyFromTexture:self.buffer sourceSlice:0 sourceLevel:0
+                sourceOrigin:MTLOriginMake((jint)(self.leftInset*self.contentsScale), (jint)(self.topInset*self.contentsScale), 0)
+                sourceSize:MTLSizeMake(self.buffer.width, self.buffer.height, 1)
+                toTexture:mtlDrawable.texture destinationSlice:0 destinationLevel:0 destinationOrigin:MTLOriginMake(0, 0, 0)];
+        [blitEncoder endEncoding];
```

Comment:
There is an issue with this code. Running Java2D.jar in Xcode asserts here with this message:

```c
2021-02-11 14:11:45.710457-0600 java[49971:9486360] Metal API Validation Enabled
2021-02-11 14:11:46.038720-0600 system_profiler[49975:9486885] Metal API Validation Enabled
-[MTLDebugBlitCommandEncoder internalValidateCopyFromTexture:sourceSlice:sourceLevel:sourceOrigin:sourceSize:toTexture:destinationSlice:destinationLevel:destinationOrigin:options:]:474: failed assertion `(sourceOrigin.y + sourceSize.height)(444) must be <= height(400).'
(lldb) 
```

and forcing the execution past it results in the java process crashing.

A solution would be to clip the src size like this:

```c
        NSUInteger src_x = self.leftInset*self.contentsScale;
        NSUInteger src_y = self.topInset*self.contentsScale;
        NSUInteger src_w = self.buffer.width-src_x;
        NSUInteger src_h = self.buffer.height-src_y;
        
        [blitEncoder
                copyFromTexture:self.buffer sourceSlice:0 sourceLevel:0
                sourceOrigin:MTLOriginMake((jint)src_x, (jint)src_y, 0)
                sourceSize:MTLSizeMake(src_w, src_h, 1)
                toTexture:mtlDrawable.texture destinationSlice:0 destinationLevel:0 destinationOrigin:MTLOriginMake(0, 0, 0)];
        [blitEncoder endEncoding];
```

Reply:
I guess you will only see this if `Metal API Validation Enabled`

Reply:
Which actually begs a question of whether we tested with `Metal API Validation Enabled` ?

Reply:
I submitted https://bugs.openjdk.java.net/browse/JDK-8261620 for this bug.
Could be that there are more such issues but since this crash occurs on start up I can't say what else there might be.

Reply:
Thanks @gerard-ziemski for bringing out this important observation. It clearly indicates that there are some overflows which are undetected.

We have tested without Metal API validation enabled till now.
 
Now, I have fixed the JDK-8261620 bug that reported the crash on app launch with Metal API validation enabled.
More testing with “Metal API validation enabled” to follow..